### ISSUE TITLE
Support CryptoConfig.CreateFromName on wasm

### ISSUE THREAD

**Issue by secana** [dotnet/runtime#44996]
### Describe the bug
In Blazor based on .NET 5, it is not possible to compute cryptographical hashes anymore. Computing a SHA256 throws the error:

```
crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: System.Security.Cryptography.Algorithms is not supported on this platform.
```

### To Reproduce

Compute a hash (SHA256, MD5, SHA1) with Blazor and .NET 5

### Impact
I'm stuck on .NET 3.1 with Blazor as my app heavily relies on cryptographical hashes. Is there a plan to add crypto support in the future?

**Comment by lewing** [comment]
@secana the hashes themselves were added back in https://github.com/dotnet/runtime/pull/40486 but none of the secure hash algorithms that require keys/secrets were permitted (HMAC etc.)

https://github.com/dotnet/runtime/pull/40486 is where we are planning how to support these scenarios in the future. What algorithms does your application require?

**Comment by secana** [comment]
I'm using the managed hashes, e.g.: 
```csharp
using var sha256 = new SHA256Managed();
var bytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(typeRefsAsString));
```

I need SHA256 and MD5 for my application to work properly. SHA1 is a nice to have but not required. Using web crypto is currently not an option, a I use the hashes in library and not directly in the Blazor code. If I use web crypto, I would need to reimplement all the business logic from the referenced library as well. 

If that helps, the library I'm talking about it [PeNet](https://github.com/secana/PeNet) where the SHA256 is used for the [TypeRefHash](https://www.gdatasoftware.com/blog/2020/06/36164-introducing-the-typerefhash-trh) and MD5 for the [ImpHash](https://www.optimizationcore.com/security/imphash-usage-malware-analysis-categorizing-malware/)

**Comment by lewing** [comment]
SHA256 should work in net5.0 browser-wasm, if it doesn't it would be helpful if you could provide a sample that fails.  We were not permitted to ship a managed MD5 implementation for multiple reasons.  If your use case requires MD5 for non cryptographic purposes you would have to fall back to your own implementation.

**Comment by GrabYourPitchforks** [comment]
Are you perhaps running a prerelease build of Blazor .NET 5? As Larry said, algorithms like SHA256 (_not_ HMACSHA256) should work just fine in that environment. But I think there were some preview builds where all primitives - including basic SHA256 - were blocked.

**Comment by secana** [comment]
Strange... I'm on NET 5.0 and the library I use is targeted at NET 5, too. Still, it does not work. I'll have a closer look later. Maybe I missed something.

**Comment by vcsjones** [comment]
> Still, it does not work.

Perhaps you are hitting the MD5 case first? MD5 is not expected to work, but SHA256 is.

**Comment by secana** [comment]
The log from the browser looks like this. 
![image](https://user-images.githubusercontent.com/12230822/101162793-12b4c180-3633-11eb-8908-a46f2c8bb4cd.png)
Seems to me that the SHA256 is the problem.

**Comment by vcsjones** [comment]
It looks to me like you're using `HashAlgorithm.Create("SHA256")` or similar based on the stack. This routes through `CryptoConfig` which I think is the part not supported.

Can you use `SHA256.Create()` instead?

**Comment by secana** [comment]
I solved the issue by implementing MD1, SHA1 and SHA256 myself in pure C#. May I ask, why the .NET 5 bcl does not shit with managed implementations of the algorithms?

**Comment by vcsjones** [comment]
> May I ask, why the .NET 5 bcl does not ship with managed implementations of the algorithms?

WASM (and only WASM) will use a managed implementation in .NET 5, with the exception of MD5, provided you access it using an API that does not go through `CryptoConfig`. You should be able to use `SHA256.Create()`, `IncrementalHash.CreateHash(..)`, and `SHA256.HashData(..)` in WASM.

I believe a future version of .NET will leverage subtle crypto in the browser and the managed implementations will be removed.

More broadly, "why isn't `SHA1Managed` 'managed'"? This is touched on a bit in the [cross-platform-cryptography][1]:

> Since the system already has a vested interest in making the cryptography libraries safe from security vulnerabilities, and already has an update mechanism that system administrators should be using, .NET gets to benefit from this reliability. Users who have requirements to use FIPS-validated algorithm implementations also get that benefit for free (when the system libraries are FIPS-validated, of course).

[1]: https://github.com/dotnet/runtime/blob/master/docs/design/features/cross-platform-cryptography.md

**Comment by akoeplinger** [comment]
It should be trivial to hardcode SHA1, SHA256 and SHA512 in `CryptoConfig.CreateFromName()`, I see no reason why we shouldn't enable that.

**Comment by GrabYourPitchforks** [comment]
"Trivial" isn't quite the word I'd use. The implementation of `CryptoConfig.CreateFromName` allows specifying friendly names ("SHA256"), fully-qualified type names, and so on. If we wanted real compat here then we should consider duplicating as much of that logic as practical. But from a security signoff perspective I think you'd be good however you wanted to approach this. Security signoff really only involves what algorithms and implementations we ship in-box. How you get there (including these APIs) is generally left up to the engineering team.

**Comment by vcsjones** [comment]
>  I see no reason why we shouldn't enable that.

There has been some effort to reduce and move people away from APIs that rely on `CryptoConfig` or using `CryptoConfig` directly, as discussed at https://github.com/dotnet/runtime/issues/37250#issuecomment-636860992.

**Comment by akoeplinger** [comment]
> The implementation of CryptoConfig.CreateFromName allows specifying friendly names ("SHA256"), fully-qualified type names, and so on.

Yup, I'd only support the mappings documented at https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptoconfig?view=net-5.0#remarks

**Comment by agnauck** [comment]
I am currently missing the following to make my app work:

- Rfc2898DeriveBytes
- HMACSHA1
- HMACSHA256
- HMACSHA384
- HMACSHA512

**Comment by GrabYourPitchforks** [comment]
@agnauck All of those (except MD5) are being tracked by https://github.com/dotnet/runtime/issues/40074. There is currently no plan to add MD5 support.

**Comment by agnauck** [comment]
whats the reasoning for not adding MD5? There is no mention about it in the other issue.

**Comment by vcsjones** [comment]
@agnauck As mentioned in #40074, the plan is to use browsers' [SubtleCrypto][1] stack to perform cryptographic operations, including hashing. SubtleCrypto in all browsers does not support MD5. 

[1]: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto

**Comment by agnauck** [comment]
@vcsjones I understand this. It would be still good to include MD5 then from managed code.
The whole purpose of NET5 and newer is having the same functionality on all supported platforms.
Right now I am catching PlatformNotSupported exceptions and then fallback to managed code included for WASM. This is not much better than the defines we had in earlier .NET version 😉

**Comment by GrabYourPitchforks** [comment]
> The whole purpose of NET5 and newer is having the same functionality on all supported platforms.

That's not 100% correct. We regularly defer to the underlying OS (or similar underlying dependency) when only the OS is able to provide a correct implementation. OS-specific features like the Registry are the most common example. But cryptography also falls into this category, because only the OS can provide a _secure_ implementation of a cryptographic algorithm. The fully-managed algorithms we ship inbox in .NET 5 (wasm only) are not approved for any security-related purposes.

The managed implementations which shipped inbox in .NET 5 wasm were the result of significant negotiation across the .NET engineering team, the Microsoft crypto board, and our corporate policy \& compliance officers. This is also contingent on using the browser-based Subtle Crypto APIs in .NET 6 wasm. Given how complex all of this was, I can only imagine how difficult it would be to retain the managed implementations of _any_ algorithm going forward, let alone a long-broken algorithm like MD5.

tl;dr: With appropriate justification we can add anything, including a managed implementation of MD5. But the bar is extraordinarily high.

**Comment by Xyncgas** [comment]
Any updates

**Comment by Xyncgas** [comment]
Mordern apps relies on security to work

**Comment by GrabYourPitchforks** [comment]
@lewing @akoeplinger et al, any reason to keep this particular issue open? We added support in .NET 5 Blazor before RTM for the core (non-keyed) hash algorithms, and https://github.com/dotnet/runtime/issues/40074 tracks adding support for the remaining algorithms.

**Comment by vcsjones** [comment]
Was there a resolution to the `CryptoConfig.CreateFromName` issue for the existing SHA hash functions, which was the cause of the initial report?

**Comment by GrabYourPitchforks** [comment]
I don't recall that being the initial cause. Somebody was instead trying to use SHA256Managed and saw that it didn't work. If we want to use this issue to track adding support for CryptoConfig, we should rename this issue to clarify that, and somebody should submit a PR.

**Comment by vcsjones** [comment]
@GrabYourPitchforks that was my understanding based on this comment. https://github.com/dotnet/runtime/issues/44996#issuecomment-738794153
